### PR TITLE
Add support for alternative axis sides in ggplotly

### PR DESCRIPTION
**Pull request by Alanocallaghan**
This is a fairly straightforward edit and shouldn't break any existing functionality. It just adds support for the new axis locations (top and right) in ggplot2 2.2.0 which already have support in `plotly.js`

### CODE REVIEW

**Comment by talgalili** [comment]
Great - thanks for your work @Alanocallaghan

**Comment by Alanocallaghan** [comment]
Cheers Tal

Addresses #808

**Comment by cpsievert** [comment]
Thanks @Alanocallaghan, this is a nice start 🍻 , but it will have to be generalized if we want it to work with facets.

To do so, I think you can essentially copy the same logic over to the [axis object generation step](https://github.com/Alanocallaghan/plotly/blob/bb7e07bd3dc15719f62b9a22d7ee6b7bf09b3e50/R/ggplotly.R#L547-L574)

And before I merge anything, it'd be great to have some tests!

**Comment by Alanocallaghan** [comment]
Thanks for getting back quickly. Moving the logic seems to work fine.

Regarding facets, some of the other new additions in ggplot 2.2.0 might also require small edits, but I've not tested any yet. Probably one to watch out for, though

Regarding tests, I've written simple tests that verify whether the position moves and ensure 1 trace

Cheers

**Comment by Alanocallaghan** [comment]
I should say that if you would like further tests or tests with more complex plots, just say the word

**Comment by cpsievert** [comment]
Nice work! More tests are always good...lets put a test with facets in there as well

**Comment by Alanocallaghan** [comment]
Unfortunately, the facet functionality technically works (http://i.imgur.com/s2PIqXA.png) but requires a bit more work...

**Comment by talgalili** [comment]
Hi @cpsievert - any chance you'll give this a look? 
Adding alternative location for the axis is a feature various people have asked for.

Thanks.

**Comment by Alanocallaghan** [comment]
I can pick this up again tonight and resolve the failing test and conflicts; it will still look awful with facets for the time being

**Comment by Alanocallaghan** [comment]
@cpsievert or any others 

The test errors are now not related to my PR (they seem to be in the master branch). I can pull in the master when the current issue is resolved.

As for facet functionality, for the x axis on the top, I'm not sure of the logic needed to pad the facet "strip" (made by `make_strip_rect`) in order to show the axis in this position. 

For the y axis, the position is a bit trickier. You need to move the axis from the leftmost plot to the rightmost plot... Not sure where the logic for this is found.

For both plots, you also need to move the axis label.

Any pointers on where to look would be great, the ggplotly code is quite dense as it's not split into many functions.

**Comment by Alanocallaghan** [comment]
The parameter is axis anchoring, eg
    layout(yaxis = list(anchor = "x2"))

**Comment by Alanocallaghan** [comment]
Made some progress, but unfortunately the "annotation" x/y axis labels are difficult to move. I think the best way forward is to extend the axis ticks, via `axisObj$ticklen`, beyond the annotations. This also handles moving the tick text.

Related:

https://github.com/ropensci/plotly/issues/510

**Comment by Alanocallaghan** [comment]
I've managed to get this to work reasonable well for 2 types of facets, shown in these images. @cpsievert, since you seem to know of edge cases (as mentioned in 510 above), do you know of any other potential trouble areas? If not, I'll write some tests and resolve any conflicts. Cheers

http://imgur.com/a/DYCDt

**Comment by Alanocallaghan** [comment]
y axis labels are not the right way up...

**Comment by Alanocallaghan** [comment]
Should be good to go once I write tests for all of the changes made

**Comment by talgalili** [comment]
Thank you for all your great work @Alanocallaghan !

**Comment by Alanocallaghan** [comment]
Tests took a while to get right (can't run them locally), should be golden now

**Comment by Alanocallaghan** [comment]
@cpsievert can you have a look whenever you have time? Thanks

**Comment by ck37** [comment]
Really interested in this patch. When I use it to move the x-axis to the top (to avoid the horizontal legend overlapping bug) the x-axis overlaps with the title. Is there any way to fix that?

Code:
```r
    p = plot_ly(ellipse, x = ~x, y = ~y, split = ~source) %>%
      layout(title = "Archeological Geochemistry",
             xaxis = list(title = names(elements)[elements == input$element1],
                          side = "top"),
             yaxis = list(title = names(elements)[elements == input$element2]),
             legend = list(orientation = "h"))
```

Screenshot: http://imgur.com/a/unwwf

**Comment by Alanocallaghan** [comment]
Good spot, here's a reprex equivalent with `mtcars`:

	p <- plot_ly(mtcars, x = ~cyl, y=~mpg, color =~gear) %>%
        layout(title = "Archeological Geochemistry",
             xaxis = list(title = "X Axis Title",
                          side = "top"),
             yaxis = list(title = "Y Axis Title"),
             legend = list(orientation = "h"),
             margin=list(t=100))

**Comment by Alanocallaghan** [comment]
This is actually related to plotly itself, while this PR only touches ggplotly

File: R/ggplotly.R [hunk 1]
```diff
@@ -587,6 +591,26 @@ gg2list <- function(p, width = NULL, height = NULL,
         title = axisTitleText,
         titlefont = text2font(axisTitle)
       )
+
+      non_default_side <- isTRUE(scales$get_scales(xy)[["position"]] != default_axis)
+      
+      ## Move axis and change anchor if necessary
+      if (has_facet(plot)) {
+        if (non_default_side) {
```
Comment (cpsievert) [inline R/ggplotly.R]: where possible, I prefer:

```
if (x && y) {
}
```

to 

```
if (x) {
  if (y) {
  }
}
```

File: R/ggplotly.R [hunk 2]
```diff
@@ -587,6 +591,26 @@ gg2list <- function(p, width = NULL, height = NULL,
...
+          if (xy == "x") {
+            ## Facet labels are always on top, I hope???
+            axisObj[["ticklen"]] <- axisObj[["ticklen"]] + 
+              (unitConvert(stripText, "pixels", type) * 2.5)
+            if (nRows > 1) {
+              axisObj[["anchor"]] <- "y"
+            }
```
Comment (cpsievert) [inline R/ggplotly.R]: What if there are multiple rows?

File: tests/testthat/test-ggplot-axis.R [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	p <- p + scale_x_continuous(position="top")
+  
+  info <- save_outputs(p, "axis_move_top")
+  expect_equal(length(info$data), 1)
+  expect_identical(info$layout$xaxis$side, "top")
+})
+
+test_that("Axis position moves to right", {
+  p <- p + scale_y_continuous(position="right")
+
```
Comment (cpsievert) [inline tests/testthat/test-ggplot-axis.R]: The axis currently isn't shown because this PR hasn't yet addressed the issue of adding space for axis ticks/text in the right margin (and removing from the left) [as is done here](https://github.com/Alanocallaghan/plotly/blob/d3527db9b5fd5198d3b0752cfddfd53c6a486333/R/ggplotly.R#L640-L644)

**Comment by Alanocallaghan** [comment]
Thanks, I'll fix these and do more extensive checks tomorrow

File: R/ggplotly.R (hunk 2 above)
Comment (Alanocallaghan) [inline R/ggplotly.R]: I think the if statement is actually redundant, since it will always need to be anchored to the first row, whether there is more than one row or not

File: tests/testthat/test-ggplot-axis.R (hunk 3 above)
Comment (Alanocallaghan) [inline tests/testthat/test-ggplot-axis.R]: Fixed, thanks for the tip

File: R/ggplotly.R (hunk 1 above)
Comment (Alanocallaghan) [inline R/ggplotly.R]: Thanks, I think I'd just thought the logic was going to be more complex

File: R/ggplotly.R [hunk 4]
```diff
@@ -587,6 +591,26 @@ gg2list <- function(p, width = NULL, height = NULL,
...
+
+      non_default_side <- isTRUE(scales$get_scales(xy)[["position"]] != default_axis)
+      
+      ## Move axis and change anchor if necessary
+      if (has_facet(plot)) {
+        if (non_default_side) {
+          if (xy == "x") {
+            ## Facet labels are always on top, I hope???
+            axisObj[["ticklen"]] <- axisObj[["ticklen"]] + 
+              (unitConvert(stripText, "pixels", type) * 2.5)
```
Comment (Alanocallaghan) [inline R/ggplotly.R]: This is a magic number. I don't think ticklen can vary dynamically so it has to be large enough to accomodate large and small window sizes. It'll be too long for small window sizes but if it's extreme the user can adjust it using `p %>% layout(xaxis=list(ticklen=10), xaxis2=list(ticklen=10), ...)`. 

I'll log an issue on the plotly.js github to see if it would be possible to have ticklen optionally be a proportion of the plot rather than strictly pixels, but otherwise it's a necessary evil

File: R/ggplotly.R [hunk 5]
```diff
@@ -599,7 +599,8 @@ gg2list <- function(p, width = NULL, height = NULL,
         if (xy == "x") {
           ## Facet labels are always on top, so add tick length to move past them
           axisObj[["ticklen"]] <- axisObj[["ticklen"]] + 
-            (unitConvert(stripText, "pixels", type) * 2.5)
+            (unitConvert(stripText, "pixels", type) * 3)
```
Comment (Alanocallaghan) [inline R/ggplotly.R]: This is a magic number. I don't think ticklen can vary dynamically so it has to be large enough to accomodate large and small window sizes. It'll be too long for small window sizes but if it's extreme the user can adjust it using `p %>% layout(xaxis=list(ticklen=10), xaxis2=list(ticklen=10), ...)`. 

I'll log an issue on the plotly.js github to see if it would be possible to have ticklen optionally be a proportion of the plot rather than strictly pixels, but otherwise it's a necessary evil

**Comment by Alanocallaghan** [comment]
The failure is due to package `semver`... Going to try re-running to see if it still fails

**Comment by talgalili** [comment]
Hi @Alanocallaghan 
Any chance this could be resolved?
(once done, heatmaply would be able to support left dendrograms with plot_method="ggplot", which would be a good thing, I believe)

Thanks.

**Comment by Alanocallaghan** [comment]
The fail on R-devel seems to be down to Rserve connection to the plotly servers. Regarding this functionality, I can't do much more while #1458 is still open so it's up to @cpsievert where to go from here, though I imagine he's busy with PlotCon around now

**Comment by talgalili** [comment]
Hi @Alanocallaghan Just to be sure - I can't seem to find issue #1458, which issue did you mean?

**Comment by Alanocallaghan** [comment]
Sorry: https://github.com/plotly/plotly.js/issues/1458

**Comment by cpsievert** [comment]
Don't worry about the test failures @Alanocallaghan. I'll have a closer look here next week

**Comment by cpsievert** [comment]
@Alanocallaghan Just FYI, #929 should supersede this, but I may use some of your tests -- #929 is a refactoring of the `labs()` framework which should fix/add support for a bunch of things. Alternative axes are coupled with that logic, so I'll handle it over there.

File: tests/testthat/test-ggplot-axis.R (hunk 3 above)
Comment (KeeonTabrizi) [reply tests/testthat/test-ggplot-axis.R]: @cpsievert Is there any workaround to get the right side y axis to render in ggplotly when using `scale_y_continuous(position="right")` from ggplot → ggplotly?

**Comment by Alanocallaghan** [comment]
Closing as superseded
